// @faustjs/next/dist/cjs/export/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FaustProvider = exports.is404 = exports.getNextStaticProps = exports.getNextServerSideProps = exports.logQueries = exports.getClient = exports.withFaust = exports.config = void 0;
var index_js_1 = require("../config/index.js");
Object.defineProperty(exports, "config", { enumerable: true, get: function () { return index_js_1.config; } });
Object.defineProperty(exports, "withFaust", { enumerable: true, get: function () { return index_js_1.withFaust; } });
var index_js_2 = require("../gqty/index.js");
Object.defineProperty(exports, "getClient", { enumerable: true, get: function () { return index_js_2.getClient; } });
var index_js_3 = require("../log/index.js");
Object.defineProperty(exports, "logQueries", { enumerable: true, get: function () { return index_js_3.logQueries; } });
var index_js_4 = require("../server/index.js");
Object.defineProperty(exports, "getNextServerSideProps", { enumerable: true, get: function () { return index_js_4.getNextServerSideProps; } });
Object.defineProperty(exports, "getNextStaticProps", { enumerable: true, get: function () { return index_js_4.getNextStaticProps; } });
Object.defineProperty(exports, "is404", { enumerable: true, get: function () { return index_js_4.is404; } });
var index_js_5 = require("../components/index.js");
Object.defineProperty(exports, "FaustProvider", { enumerable: true, get: function () { return index_js_5.FaustProvider; } });
