// gqty/Utils/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const debounce = require('./debounce.js');
const object = require('./object.js');
const cycle = require('./cycle.js');
const mergeWith = require('lodash/mergeWith.js');
const selectionsInclude = require('./selectionsInclude.js');
const promise = require('./promise.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

const mergeWith__default = /*#__PURE__*/_interopDefaultLegacy(mergeWith);

const isInteger = (v) => Number.isInteger(v);
function isEmptyObject(obj) {
  for (var _i in obj)
    return false;
  return true;
}

exports.debounce = debounce.debounce;
exports.deepAssign = object.deepAssign;
exports.get = object.get;
exports.isObject = object.isObject;
exports.isObjectWithType = object.isObjectWithType;
exports.isPlainObject = object.isPlainObject;
exports.set = object.set;
exports.decycle = cycle.decycle;
exports.retrocycle = cycle.retrocycle;
Object.defineProperty(exports, 'mergeWith', {
  enumerable: true,
  get: function () { return mergeWith__default["default"]; }
});
exports.isAnySelectionIncluded = selectionsInclude.isAnySelectionIncluded;
exports.isAnySelectionIncludedInMatrix = selectionsInclude.isAnySelectionIncludedInMatrix;
exports.isSelectionIncluded = selectionsInclude.isSelectionIncluded;
exports.toSetIfNeeded = selectionsInclude.toSetIfNeeded;
exports.LazyPromise = promise.LazyPromise;
exports.PLazy = promise.PLazy;
exports.createDeferredPromise = promise.createDeferredPromise;
exports.isEmptyObject = isEmptyObject;
exports.isInteger = isInteger;
