// gqty/Cache/dataCache.mjs
import { get, isObject, set } from '../Utils/object.mjs';
import mergeWith from 'lodash/mergeWith.js';

function createCache(normalization) {
  const cache = {};
  function getCacheFromSelection(selection, notFoundValue = void 0) {
    return normalization ? normalization.getCacheFromSelection(selection, notFoundValue, cache) : get(cache, selection.cachePath, notFoundValue);
  }
  function setCacheFromSelection(selection, value) {
    if (normalization && isObject(value))
      normalization.scanNormalizedObjects(value);
    set(cache, selection.cachePath, value);
  }
  function onObjectMergeConflict(currentValue, incomingValue) {
    const result = normalization == null ? void 0 : normalization.onObjectMergeConflict(currentValue, incomingValue);
    if (result)
      return result;
    if (Array.isArray(currentValue) && Array.isArray(incomingValue) && currentValue.length !== incomingValue.length) {
      return incomingValue;
    }
  }
  function mergeCache(data, prefix) {
    normalization == null ? void 0 : normalization.scanNormalizedObjects(data);
    mergeWith(cache, { [prefix]: data }, onObjectMergeConflict);
  }
  return {
    cache,
    getCacheFromSelection,
    setCacheFromSelection,
    mergeCache,
    normalizedCache: normalization == null ? void 0 : normalization.normalizedCache
  };
}

export { createCache };
