// @faustjs/next/dist/mjs/gqty/hooks/index.js
import { create as createAuthHook } from './useAuth.js';
import { create as createLazyQueryHook } from './useLazyQuery.js';
import { create as createMutationHook } from './useMutation.js';
import { create as createPaginatedQueryHook } from './usePaginatedQuery.js';
import { create as createQueryHook } from './useQuery.js';
import { create as createSubscriptionHook } from './useSubscription.js';
import { create as createTransactionQueryHook } from './useTransactionQuery.js';
import { create as createHydrateCacheHook } from './useHydrateCache.js';
import { create as createCategoryHook } from './useCategory.js';
import { create as createPostsHook } from './usePosts.js';
import { create as createPostHook } from './usePost.js';
import { create as createPageHook } from './usePage.js';
import { create as createPreviewHook, } from './usePreview.js';
import { create as createPreviewNodeHook } from './usePreviewNode.js';
import { create as createLoginHook } from './useLogin.js';
import { create as createLogoutHook } from './useLogout.js';
export function createHooks(useClient) {
    const useQuery = createQueryHook(useClient);
    const useMutation = createMutationHook(useClient);
    return {
        useQuery,
        useLazyQuery: createLazyQueryHook(useClient),
        useMutation,
        usePaginatedQuery: createPaginatedQueryHook(useClient),
        useSubscription: createSubscriptionHook(useClient),
        useTransactionQuery: createTransactionQueryHook(useClient),
        useHydrateCache: createHydrateCacheHook(useClient),
        useCategory: createCategoryHook(useQuery),
        usePosts: createPostsHook(useQuery),
        usePost: createPostHook(useQuery),
        usePage: createPageHook(useQuery),
    };
}
export function createAuthHooks(useClient) {
    const useQuery = createQueryHook(useClient);
    const useAuth = createAuthHook();
    const useMutation = createMutationHook(useClient);
    return {
        useQuery,
        useAuth,
        useLazyQuery: createLazyQueryHook(useClient),
        useMutation,
        usePaginatedQuery: createPaginatedQueryHook(useClient),
        useSubscription: createSubscriptionHook(useClient),
        useTransactionQuery: createTransactionQueryHook(useClient),
        useHydrateCache: createHydrateCacheHook(useClient),
        useCategory: createCategoryHook(useQuery),
        usePosts: createPostsHook(useQuery),
        usePost: createPostHook(useQuery),
        usePage: createPageHook(useQuery),
        usePreview: createPreviewHook(useAuth, useQuery),
        usePreviewNode: createPreviewNodeHook(useAuth, useQuery),
        useLogin: createLoginHook(useMutation),
        useLogout: createLogoutHook(),
    };
}
