// @faustjs/next/dist/cjs/utils/assert.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatUrlPrefix = exports.hasCategorySlug = exports.hasCategoryId = exports.hasPageUri = exports.hasPageId = exports.hasPostUri = exports.hasPostSlug = exports.hasPostId = void 0;
const isArray_js_1 = __importDefault(require("lodash/isArray.js"));
const isObject_js_1 = __importDefault(require("lodash/isObject.js"));
const isString_js_1 = __importDefault(require("lodash/isString.js"));
const trim_js_1 = __importDefault(require("lodash/trim.js"));
function hasPropString(prop) {
    return (params) => {
        return (0, isObject_js_1.default)(params) && (0, isString_js_1.default)(params[prop]);
    };
}
function hasPropStringArray(prop) {
    return (params) => {
        var _a;
        return ((0, isObject_js_1.default)(params) && (0, isArray_js_1.default)(params[prop]) && (0, isString_js_1.default)((_a = params[prop]) === null || _a === void 0 ? void 0 : _a[0]));
    };
}
exports.hasPostId = hasPropString('postId');
exports.hasPostSlug = hasPropString('postSlug');
exports.hasPostUri = hasPropStringArray('postUri');
exports.hasPageId = hasPropString('pageId');
exports.hasPageUri = hasPropStringArray('pageUri');
exports.hasCategoryId = hasPropString('categoryId');
exports.hasCategorySlug = hasPropString('categorySlug');
function formatUrlPrefix(pathPrefix) {
    if (pathPrefix.length === 1 && pathPrefix === '/') {
        return '';
    }
    const prefix = (0, trim_js_1.default)(pathPrefix, '/');
    return `/${prefix}`;
}
exports.formatUrlPrefix = formatUrlPrefix;
