// @faustjs/core/dist/cjs/config/config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getGqlUrl = exports.config = exports.normalizeConfig = exports.LOGOUT_ENDPOINT_PARTIAL_PATH = exports.TOKEN_ENDPOINT_PARTIAL_PATH = void 0;
const isString_js_1 = __importDefault(require("lodash/isString.js"));
const defaults_js_1 = __importDefault(require("lodash/defaults.js"));
const trimEnd_js_1 = __importDefault(require("lodash/trimEnd.js"));
const extend_js_1 = __importDefault(require("lodash/extend.js"));
const isObject_js_1 = __importDefault(require("lodash/isObject.js"));
const isNil_js_1 = __importDefault(require("lodash/isNil.js"));
const trim_js_1 = __importDefault(require("lodash/trim.js"));
const index_js_1 = require("../utils/index.js");
exports.TOKEN_ENDPOINT_PARTIAL_PATH = 'auth/token';
exports.LOGOUT_ENDPOINT_PARTIAL_PATH = 'auth/logout';
/* eslint-enable @typescript-eslint/ban-types */
let faustConfig = {
    wpUrl: '/',
};
let configSet = false;
/**
 * Takes a Config and ensures the properties that need to be normalized
 * (e.g. URL slashes trimmed, etc) are handled.
 *
 * @export
 * @param {Config} config
 * @returns {Config}
 */
function normalizeConfig(config) {
    const cfg = (0, defaults_js_1.default)({}, config, {
        apiBasePath: '/api/faust',
        authType: 'redirect',
        loginPagePath: '/login',
        disableLogging: false,
    });
    Object.keys(cfg).forEach((key) => {
        const keyValue = key;
        const value = cfg[keyValue];
        if ((0, isString_js_1.default)(value)) {
            cfg[keyValue] = value.trim();
        }
    });
    let { wpUrl, apiBasePath } = cfg;
    wpUrl = (0, trimEnd_js_1.default)(wpUrl, '/');
    apiBasePath = `/${(0, trim_js_1.default)(apiBasePath, '/')}`;
    return (0, extend_js_1.default)(cfg, {
        wpUrl,
        apiBasePath,
    });
}
exports.normalizeConfig = normalizeConfig;
/**
 * A setter/getter for the Config
 *
 * @export
 * @param {Config} [cfg]
 * @returns {Config}
 */
function config(cfg) {
    if (!configSet && !(0, isObject_js_1.default)(cfg)) {
        throw new Error('You must set your faustjs configuration at the highest level in your application. `config` was called with no arguments prior to setting the configuration.');
    }
    if (!(0, isObject_js_1.default)(cfg)) {
        return faustConfig;
    }
    configSet = true;
    faustConfig = normalizeConfig(cfg);
    return faustConfig;
}
exports.config = config;
/**
 * Get the full URL to the GraphQL endpoint
 *
 * @export
 * @returns
 */
function getGqlUrl() {
    const { wpUrl, gqlUrl } = config();
    if ((0, isNil_js_1.default)(gqlUrl) || !(0, isString_js_1.default)(gqlUrl)) {
        return `${wpUrl}/graphql`;
    }
    if ((0, index_js_1.isValidUrl)(gqlUrl)) {
        return (0, trimEnd_js_1.default)(gqlUrl, '/');
    }
    else {
        return `${wpUrl}/${(0, trim_js_1.default)(gqlUrl, '/')}`;
    }
}
exports.getGqlUrl = getGqlUrl;
